tests/test_transaction.py

```diff
@@ -246,39 +286,172 @@ def test_nested_inner_scope_exception_handled_in_outer_scope(conn, svcconn):
                 insert_row(conn, "inner")
                 raise ExpectedException()
         insert_row(conn, "outer-after")
-    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+    assert_not_in_transaction(conn)
     assert_rows(conn, {"outer-before", "outer-after"})
     assert_rows(svcconn, {"outer-before", "outer-after"})
 
 
 def test_nested_three_levels_successful_exit(conn, svcconn):
-    with conn.transaction():
+    """Exercise management of more than one savepoint."""
+    with conn.transaction():  # BEGIN
         insert_row(conn, "one")
-        with conn.transaction():
+        with conn.transaction():  # SAVEPOINT tx_savepoint_1
             insert_row(conn, "two")
-            with conn.transaction():
+            with conn.transaction():  # SAVEPOINT tx_savepoint_2
                 insert_row(conn, "three")
-    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+    assert_not_in_transaction(conn)
     assert_rows(conn, {"one", "two", "three"})
     assert_rows(svcconn, {"one", "two", "three"})
 
 
-def insert_row(conn, value):
-    conn.cursor().execute("INSERT INTO test_table VALUES (%s)", (value,))
+def test_named_savepoints(conn, caplog):
+    """
+    Entering a transaction context will do one of these these things:
+    1. Begin an outer transaction (if one isn't already in progress)
+    2. Begin an outer transaction and create a savepoint (if one is named)
+    3. Create a savepoint (if a transaction is already in progress)
+       either using the name provided, or auto-generating a savepoint name.
+    """
+    with caplog.at_level(logging.DEBUG, logger=transaction._log.name):
```

Comment:
Logging is not part of the adapter interface, and I wouldn't like tests to break because we change it. If you want to test that the adapter keeps on behave well you can monkeypatch the `conn._exec_command` method

Reply:
Nice. I felt a bit icky relying on logging. This is a much better approach. Thanks.

Reply:
pytest's monkeypatching should be enough for the task: https://docs.pytest.org/en/stable/monkeypatch.html

Reply:
I used `mock.patch` through force-of-habit:
https://github.com/asqui/psycopg3/pull/2/commits/8f758bb63f2615a62c868e567fa8126c9f6654fb

Any objections, given that `mock` is now in the standard library?

Reply:
Some of the unittest maintainers [have ideas I don't appreciate](https://github.com/psycopg/psycopg2/blob/cecff195fc17a83d593dd62c239aa188883a844e/tests/testutils.py#L63-L72) and I would rather stay clear of it.

I have already used pytest's monkeypatch in other places: unless it's tricky I would prefer to avoid using unittest.